src/object/cli_obj.c

```diff
@@ -2300,23 +2304,37 @@ obj_comp_cb(tse_task_t *task, void *data)
 		obj_retry_cb(task, obj, obj_auxi, pm_stale);
 
 	if (!obj_auxi->io_retry) {
-		if (obj_auxi->opc == DAOS_OBJ_RPC_SYNC &&
-		    task->dt_result != 0) {
-			struct daos_obj_sync_args	*sync_args;
+		switch (obj_auxi->opc) {
+		case DAOS_OBJ_RPC_SYNC:
+			if (task->dt_result != 0) {
+				struct daos_obj_sync_args	*sync_args;
 
-			sync_args = dc_task_get_args(task);
-			D_ASSERT(sync_args->epochs_p != NULL);
+				sync_args = dc_task_get_args(task);
+				D_ASSERT(sync_args->epochs_p != NULL);
 
-			D_FREE(*sync_args->epochs_p);
-			*sync_args->epochs_p = NULL;
-			*sync_args->nr = 0;
-		}
-		if (obj_auxi->opc == DAOS_OBJ_RPC_UPDATE ||
-		    obj_auxi->opc == DAOS_OBJ_RPC_FETCH)
+				D_FREE(*sync_args->epochs_p);
+				*sync_args->epochs_p = NULL;
+				*sync_args->nr = 0;
+			}
+			break;
+		case DAOS_OBJ_RPC_UPDATE:
+			if (!daos_handle_is_inval(obj_auxi->th))
+				dc_tx_non_cpd_cb(obj_auxi->th, task->dt_result);
+			/* Fall through. */
+		case DAOS_OBJ_RPC_FETCH:
 			/** checksums sent and not retrying,
 			 * can destroy now
 			 */
 			obj_rw_csum_destroy(obj, obj_auxi);
+			break;
+		case DAOS_OBJ_RPC_PUNCH:
+		case DAOS_OBJ_RPC_PUNCH_DKEYS:
+		case DAOS_OBJ_RPC_PUNCH_AKEYS:
+			if (!daos_handle_is_inval(obj_auxi->th))
+				dc_tx_non_cpd_cb(obj_auxi->th, task->dt_result);
+			break;
+		}
```

Comment:
we need to handle other ops, like enumerate and key_query?
If not add them then D_ASSERT(0).

Reply:
The other ops will fall through to break the switch(), not assert(0).